File: OPERATORS/LOAD_CSV/test/stress_test_load_csv_fast.lua
local ffi = require "ffi"
local qconsts = require "Q/UTILS/lua/q_consts"

ffi.cdef([[
void *memset(void *s, int c, size_t n);
void *memcpy(void *dest, const void *src, size_t n);
void * malloc(size_t size);
void free(void *ptr);
int
load_csv_fast(
    const char * const q_data_dir,
    const char * const infile,
    uint32_t nC,
    uint64_t *ptr_nR,
    const char ** fldtypes, /* [nC] */
    bool is_hdr, /* [nC] */
    bool * is_load, /* [nC] */
    bool * has_nulls, /* [nC] */
    uint64_t * num_nulls, /* [nC] */
    char ***ptr_out_files,
    char ***ptr_nil_files,
    /* Note we set nil_files and out_files only if below == NULL */
    char *str_for_lua,
    size_t sz_str_for_lua,
    int *ptr_n_str_for_lua 
    );
]])

-- TODO: Put this back later ffi.new = nil 


local function load_csv_fast_C(i)

  x = ffi.load("./load_csv_fast.so")
  local nR = ffi.cast("uint64_t *", 
             ffi.gc(
                ffi.C.malloc(ffi.sizeof("uint64_t")), ffi.C.free))
  nR[0] = 0
  nC = 5
  local fldtypes = ffi.gc(ffi.C.malloc(nC * ffi.sizeof("char *")), ffi.C.free)
  fldtypes = ffi.cast("const char **", fldtypes)
  
  local is_load = ffi.gc(ffi.C.malloc(nC * ffi.sizeof("bool")), ffi.C.free)
  is_load = ffi.cast("bool *", is_load)

  local has_nulls = ffi.gc(ffi.C.malloc(nC * ffi.sizeof("bool")), ffi.C.free)
  has_nulls = ffi.cast("bool *", has_nulls)  
  
  local num_nulls = ffi.gc(ffi.C.malloc(nC * ffi.sizeof("uint64_t")), ffi.C.free)
  num_nulls = ffi.cast("uint64_t *", num_nulls)

  local fld_name_width = 4 -- TODO Undo this hard coiding
  for i = 1, nC do
    fldtypes[i-1] = ffi.gc(ffi.C.malloc(fld_name_width * ffi.sizeof("char")), ffi.C.free)
    fldtypes[i-1] = ffi.cast("const char *", fldtypes[i-1])
    num_nulls[i-1] = 0
  end
  fldtypes[0] = "I8"
  fldtypes[1] = "SC"
  fldtypes[2] = "I4"
  fldtypes[3] = "SC"
  fldtypes[4] = "F8"

  is_load[0] = true
  is_load[1] = false
  is_load[2] = true
  is_load[3] = false
  is_load[4] = true

  has_nulls[0] = false
  has_nulls[1] = false
  has_nulls[2] = false
  has_nulls[3] = false
  has_nulls[4] = false

  local out_files = nil
  local nil_files = nil 

  local sz_str_for_lua = qconsts.sz_str_for_lua

  local str_for_lua = ffi.gc(ffi.C.malloc(sz_str_for_lua), ffi.C.free)
  str_for_lua = ffi.cast("char *", str_for_lua)
  ffi.copy(str_for_lua, "01234567890123456789012345678901234567890123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ");

  local n_str_for_lua = ffi.gc(ffi.C.malloc(ffi.sizeof("int")), ffi.C.free)
  n_str_for_lua = ffi.cast("int *", n_str_for_lua)
  n_str_for_lua[0] = 0

  data_dir = "/home/subramon/local/Q/data/"
  infile = "/home/subramon/WORK/Q/TESTS/AB1/data/eee_1.csv"

  local is_hdr = true

  local status = x.load_csv_fast(data_dir, infile, nC, nR, fldtypes,
  is_hdr, is_load, has_nulls, num_nulls, out_files, nil_files,
  str_for_lua, sz_str_for_lua, n_str_for_lua);

  --[[
  if ( ( i % 1000 ) == 1 ) then
    print(ffi.string(str_for_lua))
  end
  --]]
end

niter = 1000000
datadir = "/home/subramon/local/Q/data/"  -- FIX TODO 
command = "rm -f " .. datadir .. "/_*"
os.execute(command)
for i = 1, niter do
  load_csv_fast_C(i)
  if ( ( i % 1000 ) == 0 ) then
    print("iter = ", i)
    os.execute(command)
    collectgarbage()
  end
end
os.execute(command)
print("ALL DONE")

File: OPERATORS/LOAD_CSV/test/stress_test_load_csv_fast.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdint.h>
#include "load_csv_fast.h"
#include "_get_file_size.h"
#include "q_incs.h"
#include "_mmap.h"

#define MAX_NUM_COLS 2048
// _f1024 is 6 chars and then one space for null char
#define MAX_LEN_FILE_NAME 256
int
main(
    int argc,
    char **argv
    ) 
{
  int status = 0;
  char *fldtypes[MAX_NUM_COLS];
  bool has_nulls[MAX_NUM_COLS];
  bool is_load[MAX_NUM_COLS];
  uint64_t num_nulls[MAX_NUM_COLS];

  int niters;
  if ( argc == 2 ) {
    niters = atoi(argv[1]); 
  }
  else {
    niters = 1;
  }
  const char *data_dir = "/tmp/";
  const char *infile = "/home/subramon/WORK/Q/TESTS/AB1/data/eee_1.csv";
  uint32_t nC = 5;
  uint64_t nR = 0;
  //----------------------
  for ( int i = 0; i < MAX_NUM_COLS; i++ ) { fldtypes[i] = NULL; }
  fldtypes[0] = strdup("I8");
  fldtypes[1] = strdup("SC");
  fldtypes[2] = strdup("I4");
  fldtypes[3] = strdup("SC");
  fldtypes[4] = strdup("F8");
  //----------------------
  bool is_hdr = true;
  is_load[0] = true;
  is_load[1] = false;
  is_load[2] = true;
  is_load[3] = false;
  is_load[4] = true;
  //----------------------
  has_nulls[0] = false;
  has_nulls[1] = true;
  has_nulls[2] = true;
  has_nulls[3] = false;
  has_nulls[4] = true;
  //----------------------
  for ( int i = 0; i <  MAX_NUM_COLS; i++ ) { num_nulls[i] = 0; }
  char **out_files = NULL;
  char **nil_files = NULL;
  char *str_for_lua = NULL;
  int sz_str_for_lua = 8192;
  str_for_lua = malloc(sz_str_for_lua);
  return_if_malloc_failed(str_for_lua);
  int n_str_for_lua = 0;

  for ( int i = 0; i < niters; i++ ) { 
  status = load_csv_fast(
      data_dir, 
      infile, 
      nC, 
      &nR, 
      fldtypes, 
      is_hdr, 
      is_load, 
      has_nulls, 
      num_nulls, 
      &out_files, 
      &nil_files,
      str_for_lua, sz_str_for_lua, &n_str_for_lua);
  cBYE(status);
  if ( ( i % 10000 ) == 0 )  {
    fprintf(stderr, "Iter = %d \n", i);
    system("rm -f /tmp/_*.bin");
  }
  }
  system("rm -f /tmp/_*.bin");
  fprintf(stdout, "SUCCESS\n");
BYE:
  for ( int i = 0; i < MAX_NUM_COLS; i++ ) { 
    free_if_non_null(fldtypes[i]);
  }
  free_if_non_null(str_for_lua);
  return status;
}
